### PR TITLE
support heap sort in sql layer.

### PR DESCRIPTION
**Pull request by hanfei1991**
I will add some bench result in next commit.
@shenli @coocood @zimulala @tiancaiamao @XuHuaiyu PTAL

### CODE REVIEW

**Comment by hanfei1991** [comment]
bench result : 
old version: 	   2006591 ns/op
new version:    1313246 ns/op

**Comment by hanfei1991** [comment]
new version 65636795 ns/op
old version 68512580 ns/op

File: executor/builder.go [hunk 1]
```diff
@@ -717,12 +717,21 @@ func (b *executorBuilder) buildIndexScan(v *plan.PhysicalIndexScan, s *plan.Sele
 
 func (b *executorBuilder) buildSort(v *plan.Sort) Executor {
 	src := b.build(v.GetChildByIndex(0))
+	if v.ExecLimit != nil {
+		return &TopnExec{
```
Comment (coocood) [inline executor/builder.go]: Why not make SortExec as parent of TopnExec?

**Comment by disksing** [comment]
Why not use official `container/heap`?

**Comment by hanfei1991** [comment]
@disksing Indeed I didn't know that package ...

File: executor/builder.go (hunk 1 above)
Comment (hanfei1991) [inline executor/builder.go]: I don't really get your point.

Comment (coocood) [inline executor/builder.go]: Never mind.

File: executor/builder.go [hunk 2]
```diff
@@ -717,12 +717,21 @@ func (b *executorBuilder) buildIndexScan(v *plan.PhysicalIndexScan, s *plan.Sele
...
+			SortExec: SortExec{
+				Src:     src,
+				ByItems: v.ByItems,
+				ctx:     b.ctx,
+				schema:  v.GetSchema()},
+			limit: v.ExecLimit,
+		}
```
Comment (shenli) [inline executor/builder.go]: Can we set v.ExecLimit to nil now?

Comment (hanfei1991) [inline executor/builder.go]: Why set it to nil ? It will cause panic in topn exec

Comment (shenli) [inline executor/builder.go]: I mean set ExecLimit to nil in plan. Not in executor.

File: executor/executor.go [hunk 3]
```diff
@@ -1488,6 +1466,62 @@ func (e *SortExec) Next() (*Row, error) {
 	return row, nil
 }
 
+type TopnExec struct {
```
Comment (shenli) [inline executor/executor.go]: Add comment for this struct.

File: executor/builder.go (hunk 2 above)
Comment (hanfei1991) [inline executor/builder.go]: If execLimit is nil, we will use sort operator, else use topn operator.

File: util/heap/heap.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package heap
+
+import (
+	"sort"
+)
+
+func Heapify(heapPool sort.Interface) {
```
Comment (shenli) [inline util/heap/heap.go]: Need comments for pubic functions.

File: util/heap/heap.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		} else if rChild < curSize && cas(pos, rChild, heapPool) {
+			pos = rChild
+		} else {
+			break
+		}
+	}
+}
+
+func cas(i, j int, heapPool sort.Interface) bool {
+	if heapPool.Less(i, j) {
```
Comment (shenli) [inline util/heap/heap.go]: Will this heap support both max-heap and min-heap?

File: executor/executor.go [hunk 6]
```diff
@@ -1488,6 +1466,62 @@ func (e *SortExec) Next() (*Row, error) {
...
+	SortExec
+	limit      *plan.Limit
+	resultRows []*orderByRow
+}
+
+func (e *TopnExec) Next() (*Row, error) {
+	if !e.fetched {
+		e.Idx = 1 + int(e.limit.Offset)
+		totalCount := int(e.limit.Offset + e.limit.Count)
+		e.Rows = make([]*orderByRow, 1, totalCount+2)
```
Comment (shenli) [inline executor/executor.go]: Why use 1 for initial length?

File: executor/executor.go [hunk 7]
```diff
@@ -1488,6 +1466,62 @@ func (e *SortExec) Next() (*Row, error) {
...
 
+type TopnExec struct {
+	SortExec
+	limit      *plan.Limit
+	resultRows []*orderByRow
+}
+
+func (e *TopnExec) Next() (*Row, error) {
+	if !e.fetched {
+		e.Idx = 1 + int(e.limit.Offset)
```
Comment (shenli) [inline executor/executor.go]: Why index is 1 + offset?

Comment (hanfei1991) [inline executor/executor.go]: It's a trick: it's easy to calculate index by setting first index to 1, then i's left child must be 2 * i, right child must be 2 * i + 1, i's parent must be i / 2

File: executor/executor.go (hunk 6 above)
Comment (hanfei1991) [inline executor/executor.go]: It's a trick: it's easy to calculate index by setting first index to 1, then i's left child must be 2 * i, right child must be 2 * i + 1, i's parent must be i / 2

File: util/heap/heap.go [hunk 8]
```diff
@@ -17,6 +17,7 @@ import (
 	"sort"
 )
 
+// Heapify will put the top element to the right position in heap, which cost log2n time.
```
Comment (XuHuaiyu) [inline util/heap/heap.go]: log2n --> log2(n) be better?

File: util/heap/heap.go [hunk 9]
```diff
@@ -44,6 +45,7 @@ func cas(i, j int, heapPool sort.Interface) bool {
 	return false
 }
 
+// Update will be called when appending a element to end of heap, which costs log2n time.
```
Comment (XuHuaiyu) [inline util/heap/heap.go]: Same as above comment.

**Comment by coocood** [comment]
LGTM

**Comment by hanfei1991** [comment]
@coocood @shenli @zimulala @XuHuaiyu I use the system heap package. PTAL

File: executor/executor.go [hunk 10]
```diff
@@ -1470,15 +1470,66 @@ func (e *SortExec) Next() (*Row, error) {
...
+}
+
+func (e *TopnExec) Push(x interface{}) {
+	e.Rows = append(e.Rows, x.(*orderByRow))
+	if e.totalCount == e.heapSize {
+		if e.Less(0, e.heapSize) {
+			e.Swap(0, e.heapSize)
+			heap.Fix(e, 0)
+		}
+		e.Rows = e.Rows[:e.heapSize]
```
Comment (coocood) [inline executor/executor.go]: Why not remove the last element in `Pop` function?

File: executor/executor.go [hunk 11]
```diff
@@ -1488,6 +1466,105 @@ func (e *SortExec) Next() (*Row, error) {
...
+				row: srcRow,
+				key: make([]types.Datum, len(e.ByItems)),
+			}
+			for i, byItem := range e.ByItems {
+				orderRow.key[i], err = byItem.Expr.Eval(srcRow.Data, e.ctx)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+			}
+			heap.Push(e, orderRow)
```
Comment (coocood) [inline executor/executor.go]: Why not pop the heap if it exceeds limit count, after push?

Comment (hanfei1991) [inline executor/executor.go]: We need "update" but not "pop and push".

File: executor/executor.go [hunk 12]
```diff
@@ -1488,6 +1466,110 @@ func (e *SortExec) Next() (*Row, error) {
...
+
+// Less implements heap.Interface Less interface.
+func (e *TopnExec) Less(i, j int) bool {
+	for index, by := range e.ByItems {
+		v1 := e.Rows[i].key[index]
+		v2 := e.Rows[j].key[index]
+
+		ret, err := v1.CompareDatum(v2)
+		if err != nil {
+			e.err = err
```
Comment (shenli) [inline executor/executor.go]: Trace？

File: executor/executor.go [hunk 13]
```diff
@@ -1488,6 +1466,110 @@ func (e *SortExec) Next() (*Row, error) {
...
+				}
+				e.Rows = e.Rows[:e.heapSize]
+			} else {
+				heap.Push(e, orderRow)
+			}
+		}
+		if e.limit.Offset == 0 {
+			sort.Sort(&e.SortExec)
+		} else {
+			for i := 0; i < int(e.limit.Count) && e.Len() > 0; i++ {
```
Comment (shenli) [inline executor/executor.go]: i < e.limit.Offset?

Comment (hanfei1991) [inline executor/executor.go]: no, I should be i < e.limit.count

**Comment by coocood** [comment]
LGTM

File: executor/builder.go [hunk 14]
```diff
@@ -717,12 +717,21 @@ func (b *executorBuilder) buildIndexScan(v *plan.PhysicalIndexScan, s *plan.Sele
 
 func (b *executorBuilder) buildSort(v *plan.Sort) Executor {
 	src := b.build(v.GetChildByIndex(0))
+	if v.ExecLimit != nil {
+		return &TopnExec{
+			SortExec: SortExec{
+				Src:     src,
+				ByItems: v.ByItems,
+				ctx:     b.ctx,
+				schema:  v.GetSchema()},
```
Comment (tiancaiamao) [inline executor/builder.go]: format

**Comment by shenli** [comment]
LGTM

File: executor/builder.go (hunk 14 above)
Comment (hanfei1991) [inline executor/builder.go]: This format is also okey :)

**Pull request merged**